multivac/gather_test_data.py

```diff
@@ -0,0 +1,152 @@
+                    self.since_seconds = int(since.rstrip('h')) * 3600
+                except ValueError:
+                    print(wrong_usage_message.format('number'))
+            else:
+                print(wrong_usage_message.format('unit'))
+
+    def collect_data(self):
+        curr_time = datetime.timestamp(datetime.now())
+        print('Starting to collect data...')
+        for log in self.job_log_files:
+            log_id = log.lstrip(self.workflow_run_jobs_dir + '/').rstrip('.log')
+            self.gathered_test_data[log_id] = list()
+            with open(log, 'r') as log_file:
+                log_file_list = list(log_file)
+            test_number = 1
+
+            if self.since_seconds:
+                time = github_time_to_unix(log_file_list[0].split(
+                    ' ', 1)[0].split('.', 1)[0])
+                if curr_time - time > self.since_seconds:
+                    self.gathered_test_data.pop(log_id)
+                    break
+                    # Because list of log files is sorted
+
+            for test, conf, status in test_status_iter(log_file_list):
+                test_record = {
+                    'name': test,
+                    'configuration': conf,
+                    'status': status
+                }
+                if test_record and test_record['status'] == "fail":
+                    test_record['test_number'] = test_number
+                    test_number += 1
+                    self.gathered_test_data[log_id].append(test_record)
+
+            if len(self.gathered_test_data[log_id]):
+                with open(f'{self.workflow_run_jobs_dir}/{log_id}.json',
+                          'r') as f:
+                    job_json = json.load(f)
+                workflow_data = {
+                    'job_name': job_json['name'],
+                    'time': job_json['started_at'],
+                    'commit_sha': job_json['head_sha']
+                }
+                self.gathered_test_data[log_id].append(workflow_data)
+            else:
+                self.gathered_test_data.pop(log_id)
+        print('All test data collected successfully')
+
+    def write_to_json(self):
+        print('Writing data to a JSON file...')
+        if not os.path.isdir(self.output_dir):
+            os.makedirs(self.output_dir)
+        output_file = os.path.join(self.output_dir + '/tests.json')
+        with open(output_file, 'w') as jsonfile:
+            json.dump(self.gathered_test_data, jsonfile, indent=2)
+        print(f'Data written to {output_file}.')
+
+    def write_to_db(self):
+        influx_test_bucket = os.environ['INFLUX_TEST_BUCKET']
+        influx_org = os.environ['INFLUX_ORG']
+        data_list = []
+        print('Starting to write data to InfluxDB...')
+        for job_id in list(self.gathered_test_data.keys())[:2]:
+            if not self.gathered_test_data[job_id]:
+                continue
+            metadata = self.gathered_test_data[job_id][-1]
```

Comment:
Structure of `gather_test_data` values is unclear. What's in the last element? Could you please add an example of it's structure somewhere in the `collect_data` method?

Reply:
Oh, I see. The list contains data of each individual test and then the metadata as the last item. It's good that you only store metadata once. But it's confusing that different data structures are stored in the same list. Maybe, create another dictionary `test_metadata: dict <str, dict>`

Reply:
This way in JSON file we won't get info about workflow, because all of this info is in the `output/workflows.json` file collected by `gather_job_data.py`, did I get it right? If so, done